File: strategy/java/ParkingBill.java
import java.time.temporal.ChronoUnit;

interface Veicle {}

interface ValueCalculation {

	double calculate(long period, Veicle veicle);
}

final class DailyCalculation implements ValueCalculation {

	private final double dailyValue;

	public DailyCalculation(double dailyValue) {
		this.dailyValue = dailyValue;
	}

	@Override
	public double calculate(long period, Veicle veicle) {
		return dailyValue * Math.ceil(period / ChronoUnit.HOURS.getDuration().toSeconds());
	}
}

final class ParkingBill {

	private ValueCalculation calculation;
	private Veicle veicle;
	private long start, end;

	public ParkingBill(ValueCalculation calculation, Veicle veicle, long start, long end) {
		this.calculation = calculation;
		this.veicle = veicle;
		this.start = start;
		this.end = end;
	}

	public double valueBill() {
		return calculation.calculate(start - end, veicle);
	}

	public void setCalculation(ValueCalculation calculation) {
		this.calculation = calculation;
	}


	public static void main(String[] args) {
		Veicle car = new Veicle() { };

		ValueCalculation dailyCalculation = new DailyCalculation(34.0D);
		ParkingBill parkingBill = new ParkingBill(dailyCalculation, car, 3143894L, 2349017348L);
		parkingBill.setCalculation(dailyCalculation);

		double computedValue = parkingBill.valueBill();

		System.out.println(computedValue);
	}
}

File: strategy/kotlin/parking_bill.kt
import kotlin.time.*

interface Veicle

interface ValueCalculation {

	fun calculate(period: Long, veicle: Veicle): Double
}

@kotlin.time.ExperimentalTime
class DailyCalculation(val daily: Double) : ValueCalculation {

	override fun calculate(period: Long, veicle: Veicle): Double {
		return daily * Math.ceil(period / 1.hours.inSeconds)
	}
}

class ParkingBill(
	val veicle: Veicle,
	var calculation: ValueCalculation,
	val start: Long, 
	val end: Long
) {
	fun valueBill(): Double = calculation.calculate(end - start, veicle)
}

@kotlin.time.ExperimentalTime
fun main(): Unit {
	val car: Veicle = object : Veicle {}
	val calculation: ValueCalculation = DailyCalculation(293.90)
	val parkingBill: ParkingBill = ParkingBill(car, calculation, 3124901349L, 4123487891L)

	println(parkingBill.valueBill())
}